src/Router.php

```diff
@@ -277,18 +274,22 @@ public function clearAll(): void
     protected function invokeRoute(
         ServerRequestInterface $request,
         ResponseInterface $response,
-        RouteRunner $runner,
+        ?RouteRunner $runner = null,
         bool $catching = false
     ): ResponseInterface {
         try {
+            if ($runner === null) {
+                throw new HttpNotFound($request);
+            }
+
             return $runner->run($request);
         } catch (Throwable $e) {
             $runner =
                 $this->resolveRoute($this->getCatchers()[$e::class] ?? [], $request) ??
                 $this->resolveRoute($this->getCatchers()[Throwable::class] ?? [], $request);
```

Comment:
Hard to read. I suggest something like:

```php
$catchers = $this->getCatchers();

$exceptionCatchers = [];
if (!empty($catchers[$e::class])) {
    $exceptionCatchers = $catchers[$e::class]; 
} elseif (!empty($catchers[Throwable::class])) {
    $exceptionCatchers = $catchers[Throwable::class];
}

// avoids calling resolveRoute() because ultimately nothing will happen, nothing to resolve.
if (!empty($exceptionCatchers)) {
    $runner = $this->resolveRoute($exceptionCatchers, $request);
}
```